File: system-design/systems/01-rate-limiter/python/token_bucket.py
import math
import threading
import time
from typing import Callable


class TokenBucket:
    """令牌桶限流器：以固定速率补充令牌，每次请求取走若干，桶空则拒绝。

    - capacity   桶容量（允许的瞬时突发量）
    - refill_rate 每秒补充的令牌数（长期平均速率）
    - now        可注入的单调时钟（默认 time.monotonic），测试用假时钟推进

    并发：用 threading.Lock 保护"读-补-判-扣-写"这段复合操作。
    注意 GIL 不能替代锁——`self._tokens -= n` 跨多条字节码，多线程会交叉丢更新。
    """

    def __init__(
        self,
        capacity: float,
        refill_rate: float,
        now: Callable[[], float] = time.monotonic,
    ) -> None:
        try:
            capacity = float(capacity)
            refill_rate = float(refill_rate)
        except (TypeError, ValueError) as exc:
            raise ValueError("capacity and refill_rate must be numbers") from exc
        if not math.isfinite(capacity) or capacity <= 0:
            raise ValueError("capacity must be finite and greater than zero")
        if not math.isfinite(refill_rate) or refill_rate < 0:
            raise ValueError("refill_rate must be finite and non-negative")
        if not callable(now):
            raise ValueError("now must be callable")

        self.capacity = capacity
        self.refill_rate = refill_rate
        self._now = now
        self._tokens = capacity
        self._last = now()
        self._lock = threading.Lock()

    def allow(self, n: float = 1.0) -> bool:
        try:
            n = float(n)
        except (TypeError, ValueError):
            return False
        if not math.isfinite(n) or n <= 0:
            return False

        with self._lock:
            t = self._now()
            elapsed = max(0.0, t - self._last)
            if t > self._last:
                self._last = t
            self._tokens = min(
                self.capacity,
                self._tokens + elapsed * self.refill_rate,
            )
            if self._tokens >= n:
                self._tokens -= n
                return True
            return False

    def tokens(self) -> float:
        """返回最近一次 allow 物化后的水位，不主动补充令牌。"""
        with self._lock:
            return self._tokens

File: system-design/systems/01-rate-limiter/java/TokenBucket.java
import java.util.function.LongSupplier;

/**
 * 令牌桶限流器：以固定速率补充令牌，每次请求取走若干，桶空则拒绝。
 *
 * <p>capacity = 突发上限；refillRate = 长期平均速率（令牌/秒）。
 * 时钟通过 {@link LongSupplier}（纳秒）注入，测试用假时钟推进，不靠 sleep。
 *
 * <p>并发：方法用 {@code synchronized} 保护"读-补-判-扣-写"复合操作。
 * 也可换 {@code ReentrantLock}（可超时/可中断）或对纯计数用 {@code AtomicLong}。
 */
public class TokenBucket {
    private final double capacity;
    private final double refillRate;
    private final LongSupplier nowNanos;
    private double tokens;
    private long last;

    public TokenBucket(double capacity, double refillRate) {
        this(capacity, refillRate, System::nanoTime);
    }

    /**
     * 全参构造函数。
     *
     * @param capacity  桶容量（突发上限，令牌数）
     * @param refillRate 补充速率（令牌/秒），决定长期平均放行速率
     * @param nowNanos  时钟来源，返回纳秒级时间；便于注入假时钟做确定性测试
     */
    public TokenBucket(double capacity, double refillRate, LongSupplier nowNanos) {
        if (!Double.isFinite(capacity) || capacity <= 0) {
            throw new IllegalArgumentException("capacity must be finite and greater than zero");
        }
        if (!Double.isFinite(refillRate) || refillRate < 0) {
            throw new IllegalArgumentException("refillRate must be finite and non-negative");
        }
        if (nowNanos == null) {
            throw new IllegalArgumentException("nowNanos must not be null");
        }
        this.capacity = capacity;
        this.refillRate = refillRate;
        this.nowNanos = nowNanos;
        // 初始化时桶是满的：启动即可瞬间放行最多 capacity 个请求，
        // 用于扛住刚启动时的流量洪峰（突发），之后才按 refillRate 慢慢补充
        this.tokens = capacity;
        // 记录上次补充时间，作为后续计算时间差的基准
        this.last = nowNanos.getAsLong();
    }

    /** 便捷方法：每次尝试取走 1 个令牌。 */
    public synchronized boolean allow() {
        return allow(1);
    }

    /**
     * 尝试取走 n 个令牌。
     *
     * @param n 本次请求需要的令牌数（可为小数，便于按比例限流）
     * @return true 表示放行；false 表示令牌不足、请求被限流
     */
    public synchronized boolean allow(double n) {
        if (!Double.isFinite(n) || n <= 0) {
            return false;
        }
        // 1. 读取当前时间，与上次补充时间做差得到"流逝时间"
        long t = nowNanos.getAsLong();
        // 2. 换算成秒（纳秒 / 1e9），用于按速率补充令牌
        double elapsed = Math.max(0, ((double) t - last) / 1_000_000_000.0);
        // 3. 时钟回拨时不倒退基准，避免后续重复补充
        if (t > last) {
            last = t;
        }
        // 4. 按"速率 × 时间"补充令牌，且不超过桶容量（上限截断）
        tokens = Math.min(capacity, tokens + elapsed * refillRate);
        // 5. 令牌充足则扣减并放行，否则拒绝（不扣减，保持当前水位）
        if (tokens >= n) {
            tokens -= n;
            return true;
        }
        return false;
    }

    /** 读取最近一次 allow 物化后的令牌数；本方法不主动补充。 */
    public synchronized double tokens() {
        return tokens;
    }
}
